Google.Api.Gax.Grpc/Rest/RestChannel.cs

```diff
@@ -83,10 +87,23 @@ private async Task<ReadHttpResponseMessage> SendAsync<TRequest>(RestMethod restM
             }
             httpRequest.Headers.Add(VersionHeaderBuilder.HeaderName, RestVersion);
 
-            // TODO: How do we cancel this?
-            await AddAuthHeadersAsync(httpRequest, restMethod).ConfigureAwait(false);
+            var addAuthHeadersTask = AddAuthHeadersAsync(httpRequest, restMethod);
+            var delayTask = Task.Delay(-1, cancellationToken);
+            if (delayTask == await Task.WhenAny(addAuthHeadersTask, delayTask).ConfigureAwait(false))
+            {
+                throw new InvalidOperationException("Timeout was reached when waiting for auth headers to be added for a method {restMethod.FullName}.");
+            }
 
-            var httpResponseMessage = await _httpClient.SendAsync(httpRequest, HttpCompletionOption.ResponseContentRead, cancellationToken).ConfigureAwait(false);
+            HttpResponseMessage httpResponseMessage;
+            try
+            {
+                httpResponseMessage = await _httpClient.SendAsync(httpRequest,
+                    HttpCompletionOption.ResponseContentRead, cancellationToken).ConfigureAwait(false);
+            }
+            catch(TaskCanceledException ex)
```

Comment:
I think we can remove this catch block - just let the cancellation propagate.

Reply:
The TaskCancelledException is extremely uninformative in Core 2.1 -- does not have an inner exception or any reference to a timeout. (At later versions I think they made it wrap a TimeoutException).
For the consistency I think wrapping it into a TimeoutException would be best.

Reply:
Tiny nit: space after catch.

But more importantly, I'm not at all sure about all the uses of TimeoutException here.

I've just tried making an RPC via gRPC, and it failed with a DeadlineExceeded RpcException. I don't know whether that came from the server or whether it was generated locally... would it be reasonable to synthesize that here? (It's easy enough to do.)

@amanda-tarafa any thoughts.

Reply:
There's nowhere else where we use TimeoutException that I recall, so I don't think this would be consitent with anything we already have. In Apiary, we just [buble up the cancellation exception](https://github.com/googleapis/google-api-dotnet-client/blob/main/Src/Support/Google.Apis.Core/Http/ConfigurableMessageHandler.cs#L423) regardless of whether it's because of a timeout or something else.

I think here I'd like best to be consistent with the rest of error mapping we are doing, so +1 to throwing an RpcException with a DeadlineExceeded status.

Wether we throw inmediately after we know that the deadline is exceeded, or whether we fake a 503 response, is another thing. I' would lean towards throwing inmediately, but then I think we should modify the GetStatus method (line 145) to take into account the case in which the task is faulted with an RpcException, so we can return the Status contained by that RpcException itself.

Reply:
We use TimeoutException in Polling, but only there.

I *thought* we had cunning code in RetryAttempt to avoid sleeping if we know that by the time we've finished sleeping the deadline will have passed, but I can't find that now.

Reply:
RpcException does not allow wrapping an inner exception, so we either throw two different exception types: RPC+TaskCancelled (a very uninformative TaskCancelled at that) or throw the TaskCancelled away.

Reply:
Hmm. We do have a slight issue here - while it's tempting to throw away the TaskCanceledException, if this is *actually* because the caller passed in a CancellationToken and that got canceled, it's weird to change that.

What we could potentially do is use the deadline and current time to work out what to throw - that's pretty weird, but would probably give the right expected semantics.

Reply:
There is no `caller`, only us. We generate a token from a source we control. The httpClient.SendAsync docs say that TaskCancelled exception happens because `The request failed due to timeout.`
I would throw an RpcException here, then we can ask the Grpc team to give us a constructor that takes an InnerException, and update the code to wrap a TaskCancelled later.

Reply:
I'd argue that this *is* a gRPC context, in terms of the abstraction GAX exposes - we're implementing a gRPC CallInvoker. So it makes sense to talk in gRPC concepts - just like we translate HTTP status codes into gRPC ones.

Even though we generate one CancellationToken, there's another which may be present in the CallOptions... it looks like we're not observing that at the moment, which also needs to be fixed. (The TODO was meant to cover all aspects of CallOptions; a deadline is only part of that.) But we can probably differentiate between the cancellation token *we've* created (just for the deadline) and one that has been provided by the caller. We can check whether our deadline-oriented cancellation token has been cancelled, and translate that into an RpcException, but leave anything else.

Reply:
Updated to account for this